File: frontend/src/lib/wizard.ts
/**
 * M2/M3/M4 — เรียก API ของระบบนำทาง
 *
 * ชนิดข้อมูลในไฟล์นี้ต้องตรงกับ backend/app/schemas/wizard.py
 * ถ้าแก้ฝั่งใดฝั่งหนึ่งต้องแก้อีกฝั่งด้วย
 *
 * endpoint กลุ่มนี้ไม่ต้องล็อกอิน ตั้งใจให้ลองประเมินได้ทันทีตาม US-01
 */
import { api } from "@/lib/api";

export type LocalAuthority = {
  id: number;
  code: string;
  name: string;
  kind: string;
  district: string;
};

export type ContactPoint = {
  agency_name: string;
  local_authority_name: string | null;
  office_name: string;
  address: string | null;
  phone: string | null;
  office_hours: string | null;
  estimated_days: number | null;
  notes: string | null;
};

export type RequiredDocument = {
  code: string;
  name_th: string;
  description: string | null;
  is_mandatory: boolean;
  /** true = กรอกในระบบ ไม่ต้องอัปโหลดไฟล์ (เช่น A01 แบบหนังสือแจ้งฯ) */
  is_system_form: boolean;
  allows_multiple: boolean;
  accepted_mime: string[];
  preparation_note: string | null;
  estimated_days: number | null;
  contact_point: ContactPoint | null;
};

export type ClassifyResult = {
  property_type_code: string;
  property_type_name: string;
  requires_license: boolean;
  is_out_of_scope: boolean;
  /** M2 บังคับ: เหตุผลว่าทำไมจึงได้ผลนั้น */
  reason: string;
  outcome_message: string;
  /** กฎใน DB ที่ใช้ตัดสิน — ใช้สาธิต US-09 */
  matched_rule_code: string | null;
  fee: { amount: number; currency: string; validity_years: number } | null;
  documents: {
    self_service: RequiredDocument[];
    external: RequiredDocument[];
    needs_local_authority: boolean;
  };
};

export type WizardAnswers = {
  rooms: number;
  guests: number;
  has_restaurant: boolean;
  local_authority_id: number | null;
};

export function listLocalAuthorities() {
  return api<LocalAuthority[]>("/wizard/local-authorities");
}

export function classify(answers: WizardAnswers) {
  return api<ClassifyResult>("/wizard/classify", {
    method: "POST",
    body: JSON.stringify(answers),
  });
}

/** แปลง mime เป็นคำที่ผู้ใช้ทั่วไปเข้าใจ — NFR Usability ห้ามโชว์ "application/pdf" ดิบ */
export function describeAccepted(mimes: string[]): string {
  const hasPdf = mimes.includes("application/pdf");
  const hasImage = mimes.some((m) => m.startsWith("image/"));

  if (hasPdf && hasImage) return "ไฟล์ PDF หรือรูปถ่าย";
  if (hasPdf) return "ไฟล์ PDF เท่านั้น";
  if (hasImage) return "รูปถ่าย (JPG หรือ PNG)";
  return mimes.join(", ");
}

File: backend/app/api/v1/wizard.py
"""M2 — ระบบนำทางแบบถามตอบทีละขั้น (Guided Wizard)

หัวใจของโจทย์: รับคำตอบ -> จำแนกประเภทตาม classification_rule ใน DB
-> ตอบกลับพร้อม "เหตุผลว่าทำไมจึงได้ผลนั้น" + รายการเอกสาร 2 หมวด

ข้อควรระวัง (กรณีทดสอบ T-02): เงื่อนไข "ไม่เข้าข่ายโรงแรม" คือ
ห้องพัก <= 8 ห้อง **และ** ผู้เข้าพัก <= 30 คน ต้องเป็น AND ทั้งสองเงื่อนไข
ที่พัก 8 ห้องแต่รับ 36 คน ต้องตอบว่า "เข้าข่ายต้องขอใบอนุญาต"

endpoint กลุ่มนี้ไม่บังคับให้เข้าสู่ระบบ ตั้งใจให้ผู้ประกอบการลองเช็ก
ได้ทันทีตาม US-01 ("ตอบคำถามไม่กี่ข้อแล้วรู้ว่าต้องขอใบอนุญาตหรือไม่")
และไม่เขียนอะไรลง DB จึงไม่มีขยะค้างจากคนที่แค่มาลอง

เจ้าของงานส่วนนี้: <ใส่ชื่อสมาชิก>
"""

from fastapi import APIRouter, HTTPException, status

from app.api import presenters
from app.api.deps import DbSession
from app.schemas.wizard import ClassifyResult, FeeOut, LocalAuthorityOut, WizardAnswers
from app.services import classification as svc

router = APIRouter()


@router.get(
    "/local-authorities",
    response_model=list[LocalAuthorityOut],
    summary="รายชื่อ อปท. ทั้งหมดสำหรับให้เลือกใน wizard",
)
def list_local_authorities(db: DbSession) -> list[LocalAuthorityOut]:
    return [LocalAuthorityOut.model_validate(a) for a in svc.local_authorities(db)]


@router.post(
    "/classify",
    response_model=ClassifyResult,
    summary="ประเมินว่าที่พักเข้าข่ายประเภทใด พร้อมเหตุผลและรายการเอกสาร",
    responses={
        422: {"description": "ไม่มีกฎใดในระบบรองรับกรณีนี้"},
    },
)
def classify(payload: WizardAnswers, db: DbSession) -> ClassifyResult:
    outcome = svc.classify(db, svc.Answers(payload.rooms, payload.guests, payload.has_restaurant))

    if outcome is None:
        # เกิดได้เมื่อ Super Admin แก้กฎจนมีช่องโหว่ — ต้องบอกให้ผู้ใช้เข้าใจ
        # ไม่ใช่ปล่อย 500 ออกไป (NFR Usability)
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=("ระบบยังไม่มีเกณฑ์สำหรับที่พักลักษณะนี้ กรุณาติดต่อเจ้าหน้าที่เพื่อตรวจสอบเป็นรายกรณี"),
        )

    docs = svc.required_documents(
        db, outcome.property_type.id, local_authority_id=payload.local_authority_id
    )

    return ClassifyResult(
        property_type_code=outcome.property_type.code,
        property_type_name=outcome.property_type.name_th,
        requires_license=outcome.property_type.requires_license,
        is_out_of_scope=outcome.property_type.is_out_of_scope,
        reason=outcome.reason,
        outcome_message=outcome.outcome_message,
        matched_rule_code=outcome.matched_rule.code if outcome.matched_rule else None,
        fee=FeeOut(**vars(outcome.fee)) if outcome.fee else None,
        documents=presenters.to_checklist(docs, local_authority_id=payload.local_authority_id),
    )
